lib/virt.py

```diff
@@ -820,6 +827,51 @@ def ssh_keygen(path):
     subprocess.run(['ssh-keygen', '-N', '', '-f', path], stdout=DEVNULL, check=True)
 
 
+def translate_ssg_kickstart(profile):
+    """
+    Parse (and tweak) a kickstart shipped with the upstream content
+    into class Kickstart instance.
+    """
+    ks_text = ''
+    with open(util.get_kickstart(profile)) as f:
+        for line in f:
+            line = line.rstrip('\n')
+
+            # use our own password
+            if re.match('^rootpw ', line):
+                line = f'rootpw {GUEST_LOGIN_PASS}'
+
+            # don't hardcode interface name
+            if re.match('^network ', line):
+                line = re.sub(' --device[^ ]+', '', line)
+
+            # shrink some unnecessary partitions
+            if re.match('^logvol / ', line):
+                line = re.sub('--size=[^ ]+', '--size=4000', line)
+            if re.match('^logvol /var/log/audit ', line):
+                line = re.sub('--size=[^ ]+', '--size=500', line)
+
+            ks_text += f'{line}\n'
+
+    # remove %addon oscap, we'll add our own
+    ks_text = re.sub(r'\n%addon .+?_oscap\n.+?\n%end[^\n]*', '', ks_text, flags=re.DOTALL)
+
+    # remove %packages, until it becomes consistent everywhere
+    # TODO: upstream content PR + add it to README
+    ks_text = re.sub(r'\n%packages.*?\n%end[^\n]*', '', ks_text, flags=re.DOTALL)
+
+    return Kickstart(template=ks_text)
+
+
+def firewalld_allow_port(port):
```

Comment:
Wouldn't `utils.py` be a better fit? To move it closer to HTTP Server as those two functions are related (and most likely will be the only ones).
Maybe even use within HTTP Server? Just food for thought, opening/closing the ports with HTTP server.

Reply:
I did consider that, but because it explicitly involved the `libvirt` zone, I put it in `virt.py`. I also considered putting it into the HTTP server class, but once again decided against it because of the `libvirt` zone.

Another thought I had that we could do is to generalize it as "add the port to all firewalld zones" or even "check for iptables/nftables and add a transient rule manually", ignoring firewalld and its zones, making it generic (non-libvirt), allowing it to be placed in the util/HTTP code.

Since you pointed this out, I think I'll do it. I also don't want any more code in `virt.py` than necessary.

Reply:
Hmm, perhaps that wouldn't do it.

`iptables` could simply insert an `ACCEPT` rule into `INPUT`, but `nftables` is much more complex due to its nature of multiple tables, each having multiple chains, and `accept`-ing in a higher priority one doesn't fully accept the packet, it just makes it go onto other filters.  
So we would need to insert it into every chain of every table.

At that point, we might as well just disable the firewall and restore it after the HTTP server shuts down. But that would mean using blocks of empty code for `iptables-restore` and storing the original `nft` ruleset, maybe breaking `libvirt` itself.

... So I may go the simpler way and just use libvirt's `nwfilter`, which is basically a XML language for writing firewall rules for virtual machines, and make all this libvirt-specific anyway. At least there won't be arbitrary `firewalld` stuff in `virt.py`.

Reply:
Well, that way also isn't too great, as nwfilter is a whole new socket/service that's disabled by default, and the language would need another XML block in `virt.py`.

I think I'll just do it on the firewalld level, "list all zones, parse it, add rule to every zone, prepare cleanup removing them".